salt/tests/unit/modules/test_metalk8s.py

```diff
@@ -0,0 +1,272 @@
+        ({'node1': {'roles': ['master', 'node']}, 'node2': {'roles': ['node']}}, 'node', ['node1', 'node2']),
+        ({}, 'master', []),
+        ({'node1': {}, 'node2': {}}, 'master', []),
+        ({'node1': {'roles': ['node']}, 'node2': {'roles': ['node']}}, 'master', []),
+        (None, 'master', ['node1'], False, {'node1': {'roles': ['master']}}),
+        (None, 'master', "Can't retrieve 'metalk8s:nodes' pillar: 'nodes'", True),
+        (None, 'master', "Can't retrieve minions by role because of errors in pillar 'metalk8s:nodes': Error in pillar", True, {'_errors': ['Error in pillar']})
+    ])
+    def test_minions_by_role(self, nodes, role, result,
+                             raises=False, pillar_nodes=None):
+        """
+        Tests the return of `minions_by_role` function
+        """
+        pillar_dict = {
+            'metalk8s': {}
+        }
+        if pillar_nodes:
+            pillar_dict['metalk8s']['nodes'] = pillar_nodes
+
+        with patch.dict(metalk8s.__pillar__, pillar_dict):
+            if raises:
+                self.assertRaisesRegexp(
+                    CommandExecutionError,
+                    result,
+                    metalk8s.minions_by_role,
+                    role,
+                    nodes
+                )
+            else:
+                self.assertEqual(
+                    set(metalk8s.minions_by_role(role, nodes)),
+                    set(result)
+                )
+
+    @parameterized.expand([
+        (PRODUCT_TXT, {'version': '2.5.0', 'name': 'MetalK8s'}),
+        (None, 'Path /my/path/ has no "product.txt"', True),
+        ("Not a good product txt", {'version': None, 'name': None})
+    ])
+    def test_archive_info_from_tree(self, product, result, raises=False):
+        """
+        Tests the return of `archive_info_from_tree` function
+        """
+        is_file_mock = MagicMock(return_value=product is not None)
+        open_file_mock = MagicMock()
+        read_mock = open_file_mock.return_value.__enter__.return_value.read
+        read_mock.return_value = product
+
+        with patch("os.path.isfile", is_file_mock), \
+                patch("salt.utils.files.fopen", open_file_mock):
+            if raises:
+                self.assertRaisesRegexp(
+                    CommandExecutionError,
+                    result,
+                    metalk8s.archive_info_from_tree,
+                    "/my/path/"
+                )
+            else:
+                self.assertEqual(
+                    metalk8s.archive_info_from_tree("/my/path/"),
+                    result
+                )
+                open_file_mock.assert_called_once_with("/my/path/product.txt")
+                read_mock.assert_called_once()
+
+            is_file_mock.assert_called_once_with("/my/path/product.txt")
+
+    @parameterized.expand([
+        (PRODUCT_TXT, {'version': '2.5.0', 'name': 'MetalK8s'}),
+        (None, 'Failed to run isoinfo: Error occured', True),
```

Comment:
`occurred` takes 2 R:
```suggestion
        (None, 'Failed to run isoinfo: An error has occurred', True),
```

Reply:
Yes sure that why it's an error :sweat_smile: (right I will fix it)